tmp-0.6/spark-agent0.6.MD

```diff
@@ -0,0 +1,228 @@
+- `IGNORE_LINEAGE` (default)
+
+`spline.postprocessingFilter.classNames`
+ - Optional comma separated list of user lineage filters. More info in **Post Processing Lineage Filters** chapter.
+
+## Lineage Dispatchers
+Lineage Dispatchers is a class that is usually (but not always) responsible for sending the converted lineage to gateway. By default `HttpLineageDispatcher` is used.
+
+Available dispacthers:
+- `HttpLineageDispatcher` - sends the lineage via http 
+- `KafkaLineageDispatcher` - sends the lineage via kafka
+- `ConsoleLineageDispatcher` - write the lineage to console
+- `LoggingLineageDispatcher` - logs the lineahge using logger
+- `CompositeLineageDispatcher` - allows combining multiple dispatchers
+
+Each dispatcher can have different configuration parameters. To make the configs clearly separated each dispacther has it's own namespace in which all it's parameters are defined. I will explain it on an kafka examples.
+
+Defining dispacther
+```
+spline.lineageDispatcher=kafka
+```
+Once you defined the dispatcher all other parameters will have a namespace `spline.lineageDispatcher.{{dipatcher-name}}.` as a prefix. In this case it is `spline.lineageDispatcher.kafka.`. 
+
+To find out which parameters you can use look into `spline.default.properties`. For kafka I would have to define at least these two properties:
+```
+spline.lineageDispatcher.kafka.topic=foo
+spline.lineageDispatcher.kafka.producer.bootstrap.servers=localhost:9092
+```
+
+### Creating your own dispatcher
+There is also possibility to create your own dispatcher. It must implement `LineageDispatcher` trait and have a constructor with a single parameter of type `org.apache.commons.configuration.Configuration`. To use it you must define name and class and also all other parameters you need. For example:
+```
+spline.lineageDispatcher=my-dispatcher
+spline.lineageDispatcher.my-dispatcher.className=org.example.spline.MyDispatcherImpl
+spline.lineageDispatcher.my-dispatcher.prop1=value1
+spline.lineageDispatcher.my-dispatcher.prop2=value2
+```
+
+## Post Processing Lineage Filters
+Filters can be used to enrich the lineage with your own custom data or to remove unwanted data like passwords. All filters are aplied after the Spark plan is converted to Spline DTOs, but before the dispatcher is called.
+
+The filter must implement `LineageFilter` class and have a constructor with a single parameter of type `org.apache.commons.configuration.Configuration`. For an example you can take a look at `DataSourceUriLineageFilter`. You can use the Configuration instance to acces configuration params and make the filter configurable that way. `LineageFilter`'s methods will provide you not only with lineage DTOs, but also wtih `HarvestingContext` object containing the original Spark plan.
+
+To use the filter you must set it via a property. For example:
+```
+spline.postprocessingFilter.classNames=za.co.absa.spline.harvester.postprocessing.DataSourceUriLineageFilter
+```
+
```

Comment:
``````suggestion
The procedure how filters are registered and configured is similar to the `LineageDispatcher` registration and configuration procedure.
A custom filter class must implement `za.co.absa.spline.harvester.postprocessing.PostProcessingFilter` trait and declare a constructor with a single parameter of type `org.apache.commons.configuration.Configuration`.
Then register and configure it like this: 
```properties
spline.postProcessingFilter.className=my-filter
spline.postProcessingFilter.my-filter.className=my.awesome.CustomFilter
spline.postProcessingFilter.my-filter.prop1=value1
spline.postProcessingFilter.my-filter.prop2=value2
```

Use pre-registered `CompositePostProcessingFilter` to chain up multiple filters:
```properties
spline.postProcessingFilter=composite
spline.postProcessingFilter.composite.filters=myFilter1,myFilter2
```

(see `spline.default.properties` for details and examples)
``````